src/vm/costs/mod.rs

```diff
@@ -134,6 +211,73 @@ impl LimitedCostTracker {
     }
 }
 
+fn parse_cost(cost_function: ClarityCostFunction, eval_result: InterpreterResult<Option<Value>>) -> Result<ExecutionCost> {
+    match eval_result {
+        Ok(Some(Value::Tuple(data))) => {
+            let results = (
+                data.data_map.get("write_length"),
+                data.data_map.get("write_count"),
+                data.data_map.get("runtime"),
+                data.data_map.get("read_length"),
+                data.data_map.get("read_count"),
+            );
+
+            match results {
+                (Some(UInt(write_length)),
+                    Some(UInt(write_count)),
+                    Some(UInt(runtime)),
+                    Some(UInt(read_length)),
+                    Some(UInt(read_count))) => {
+                    Ok(ExecutionCost {
+                        write_length: (*write_length as u64),
+                        write_count: (*write_count as u64),
+                        runtime: (*runtime as u64),
+                        read_length: (*read_length as u64),
+                        read_count: (*read_count as u64),
+                    })
+                },
+                _ => Err(CostErrors::CostComputationFailed("Execution Cost tuple does not contain only UInts".to_string()))
+            }
+        },
+        Ok(Some(_)) => Err(CostErrors::CostComputationFailed("Clarity cost function returned something other than a Cost tuple".to_string())),
+        Ok(None) => Err(CostErrors::CostComputationFailed("Clarity cost function returned nothing".to_string())),
+        Err(e) => Err(CostErrors::CostComputationFailed(format!("Error evaluating result of cost function {}: {}", cost_function.get_name(), e))),
+    }
+}
+
+fn compute_cost(
+    cost_tracker: &mut LimitedCostTracker,
+    cost_function: ClarityCostFunction,
+    input_size: u64) -> Result<ExecutionCost> {
+
+    let mut null_store = NullBackingStore::new();
+    let conn = null_store.as_clarity_db();
+    let mut global_context = GlobalContext::new(conn, LimitedCostTracker::new_free());
+
+    let cost_function_reference = cost_tracker.cost_function_references
```

Comment:
Is this actually a reference?  Looks like you're making an owned copy of the function here.

Reply:
This isn't actually a reference (in the Rust sense). Perhaps this is just poor naming, but I was using the word reference to mean contract ID + function name. Will reconsider this naming...

Reply:
Well, before renaming, is there a way you can reduce the number of `.clone()` calls here?  It might be fine to literally take the reference.

Reply:
I don't think it's likely to be able to convert these to non-owned references. It may be possible in this PR, but looking ahead, these references to contracts are going to be controlled by the voting contract, meaning that state will come from the Stacks chain state. So if a reference was held to that data, Stacks chain state processing could not occur as long as that reference is held (because no mutable references would be usable).

Reply:
Also -- regarding the naming, these are clarity function references: they refer to clarity functions, they are not themselves the data of clarity functions.